### PR TITLE
Storage workflow

### LINKED ISSUE #1388: new odo workflow: odo storage
odo storage should only modify the config file. Storage will be created by `odo push`

- changes to odo storage command
- add storage suppport to `odo push`

`odo storage create`
  - it just updates the `.odo/config.yml` file ( no action on the cluster)
  - `odo push` has to be run to actually create the storage on the cluster

`odo storage delete`
  - it just updates the `.odo/config.yml` file ( no action on the cluster)
  - `odo push` has to be run to actually create the storage on the cluster

`odo storage list`
  - lists all the storage that is in the config file
  - Ideally, it should list all storage associated with the component, both in the cluster and in the config file, and mark what is from the config and what is from the cluster. Let's start with listing just the config, we can add cluster side later. 
  
`odo storage mount`
 - let's skip this command for now - comment it out, we will get back to this later

`odo storage unmount`
 - let's skip this command for now - comment it out, we will get back to this later

### PR DESCRIPTION
**Pull request by mik-dass**
fixes #1388

### CODE REVIEW

**Comment by kadel** [comment]
@mik-dass this PR is growing to be too big :-( It might be good to stop at some point and address additional issues in followup PR

**Comment by mik-dass** [comment]
> @mik-dass this PR is growing to be too big :-( It might be good to stop at some point and address additional issues in followup PR

The PR is almost complete. I am just adding some e2e tests

**Comment by amitkrout** [comment]
Flake - https://github.com/openshift/odo/issues/1628

//test e2e-core

**Comment by mohammedzee1000** [comment]
```
go test -v github.com/openshift/odo/tests/e2e --ginkgo.focus="odoe2e" -ginkgo.succinct -timeout 30m
=== RUN   TestOdo
[1555920464] odo test suite - 28/93 specs SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS•••Running command: odo create nodejs
 ✗  Unable to connect to OpenShift cluster, is it down?
Result: 
 stdout: 
 stderr: ✗  Unable to connect to OpenShift cluster, is it down?
 
 exitcode: 1 

```
/retest

**Comment by mohammedzee1000** [comment]
```
level=error msg="Terraform does not automatically rollback in the face of errors."
level=error msg="Instead, your Terraform state file has been partially updated with"
level=error msg="any resources that successfully completed. Please address the error"
level=error msg="above and apply again to incrementally change your infrastructure."
level=error
level=error
level=fatal msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed to apply using Terraform"
```
/retest

**Comment by mohammedzee1000** [comment]
```
Run `odo push --config` command to apply changes to the cluster.

 stderr: 
 exitcode: 0 
Running command: odo config view
COMPONENT SETTINGS
------------------------------------------------
PARAMETER         CURRENT_VALUE
Type              java
Application       app
Project           ci-op-fw5kgzws
SourceType        local
Ref               
SourceLocation    ./
Ports             
Name              nodejs-ydxc
MinMemory         
MaxMemory         
Ignore            
MinCPU            
MaxCPU            
Result: 
 stdout: COMPONENT SETTINGS
------------------------------------------------
PARAMETER         CURRENT_VALUE
Type              java
Application       app
Project           ci-op-fw5kgzws
SourceType        local
Ref               
SourceLocation    ./
Ports             
Name              nodejs-ydxc
MinMemory         
MaxMemory         
Ignore            
MinCPU            
MaxCPU            

 stderr: 
 exitcode: 0 
Running command: odo config unset Type -f
 ✗  Please log in to the cluster
Result: 
 stdout: 
 stderr: ✗  Please log in to the cluster
 
 exitcode: 1 

------------------------------
• Failure [1.117 seconds]
odoe2e odo config [It] should be checking to see if local config values are the same as the configured ones 
/go/src/github.com/openshift/odo/tests/e2e/e2e_test.go:225

```
/retest

**Comment by mohammedzee1000** [comment]
```
To switch to this project and start adding applications, use:

    oc project grxjnqycxk

 stderr: 
 exitcode: 0 
Created dir: /tmp/574979545
Current working dir: /go/src/github.com/openshift/odo/tests/e2e
Setting current dir to: /tmp/574979545
Running command: odo component create nodejs nodejs --app nodejs-push-test --project grxjnqycxk
 ✗  this directory already contains a component
Result: 
 stdout: 
 stderr: ✗  this directory already contains a component
 
 exitcode: 1 
Setting current dir to: /go/src/github.com/openshift/odo/tests/e2e
Deleting project: grxjnqycxk
Running command: oc delete project grxjnqycxk --now
project.project.openshift.io "grxjnqycxk" deleted
Result: 
 stdout: project.project.openshift.io "grxjnqycxk" deleted

 stderr: 
 exitcode: 0 
Deleating dir: /tmp/574979545

------------------------------
• Failure [2.040 seconds]
odoCmpE2e Component tests Test odo push with --source and --config flags when project flag(--project) is used when using current directory [It] create local nodejs component and push source and code separately 
/go/src/github.com/openshift/odo/tests/e2e/component.go:553

  Expected
      <int>: 1
  to equal
      <int>: 0

  /go/src/github.com/openshift/odo/tests/e2e/helper/helper_run.go:38
```
Looks like improper cleanup

File: pkg/config/storage_test.go [hunk 1]
```diff
@@ -0,0 +1,495 @@
+package config
+
+import (
+	"io/ioutil"
+	"os"
+	"reflect"
+	"testing"
+)
+
+func TestLocalConfigInfo_StorageCreate(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/config/storage_test.go]: They are a part of the localConfigInfo, so this way we are indicating the same

File: pkg/config/config.go [hunk 2]
```diff
@@ -287,7 +301,21 @@ func (lci *LocalConfigInfo) DeleteUrl(parameter string) error {
 		}
 	}
 	return lci.writeToFile()
+}
 
+func (lci *LocalConfigInfo) DeleteFromConfigurationList(parameter string, value string) error {
```
[the remark this answers is not in the document]
Comment (mik-dass) [reply pkg/config/config.go]: Done

**Comment by surajnarwade** [comment]
travis is failing :

```
go test -v github.com/openshift/odo/tests/integration --ginkgo.focus="odoSourceE2e" -ginkgo.slowSpecThreshold=120 -ginkgo.v -timeout 1800s
# github.com/openshift/odo/tests/integration [github.com/openshift/odo/tests/integration.test]
tests/integration/storgae_test.go:46:37: cannot call pointer method on helper.NewOcRunner("oc")
tests/integration/storgae_test.go:46:37: cannot take the address of helper.NewOcRunner("oc")
FAIL	github.com/openshift/odo/tests/integration [build failed]
make: *** [test-source-e2e] Error 2
/home/travis/.travis/functions: line 524: 15630 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
The command "travis_wait make test-source-e2e" exited with 2.
```

**Comment by mik-dass** [comment]
@surajnarwade @kadel Rebased

**Comment by mik-dass** [comment]
```
level=warning msg="Found override for ReleaseImage. Please be warned, this is not advised"
level=info msg="Consuming \"Install Config\" from target directory"
level=info msg="Creating infrastructure resources..."
level=error
level=error msg="Error: Error applying plan:"
level=error
level=error msg="1 error occurred:"
level=error msg="\t* module.dns.aws_route53_record.api_external: 1 error occurred:"
level=error msg="\t* aws_route53_record.api_external: timeout while waiting for state to become 'INSYNC' (last state: 'PENDING', timeout: 30m0s)"
level=error
level=error
level=error
level=error
level=error
level=error msg="Terraform does not automatically rollback in the face of errors."
level=error msg="Instead, your Terraform state file has been partially updated with"
level=error msg="any resources that successfully completed. Please address the error"
level=error msg="above and apply again to incrementally change your infrastructure."
level=error
level=error
level=fatal msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed to apply using Terraform"
```
/retest

**Comment by mik-dass** [comment]
```
[oc] Now using project "new-project" on server "https://api.ci-op-d6hmq3ml-5186a.origin-ci-int-aws.dev.rhcloud.com:6443".
Running odo with args: [project list] 
[odo] ACTIVE     NAME
[odo]            myproject
[odo] *          new-project
Deleting project: new-project
Running odo with args: [project delete new-project -f] 
[odo]  ✗  Unable to connect to OpenShift cluster, is it down?

• Failure [142.215 seconds]
odoCmpE2e
/go/src/github.com/openshift/odo/tests/integration/cmp_test.go:13
  Creating Component even in new project
  /go/src/github.com/openshift/odo/tests/integration/component.go:262
    should create component [It]
    /go/src/github.com/openshift/odo/tests/integration/component.go:270

    No future change is possible.  Bailing out early after 1.101s.
    Expected
        <int>: 1
    to match exit code:
        <int>: 0
```
/retest

File: .travis.yml [hunk 3]
```diff
@@ -85,6 +85,16 @@ jobs:
...
 
+    - <<: *base-test
+      stage: test
+      name: "Storage e2e tests"
+      script:
+        - ./scripts/oc-cluster.sh
+        - make bin
+        - sudo cp odo /usr/bin
+        - oc login -u developer
+        - make test-storage-e2e
```
Comment (kadel) [inline .travis.yml]: It might be better to group this together with some other existing tests. Like for example watch test

**Comment by mik-dass** [comment]
@kadel Fixed.

**Comment by mik-dass** [comment]
```
Failed to login as developer
make: *** [configure-installer-tests-cluster] Error 1
```
/retest

**Pull request merged**